golang/Chapter06/model/client.go

```diff
@@ -0,0 +1,755 @@
+	if res < limit {
+		return identifier
+	}
+
+	pipeline.ZRem(semname, identifier)
+	pipeline.ZRem(czset, identifier)
+	if _, err := pipeline.Exec(); err != nil {
+		log.Println("pipeline err in AcquireFairSemaphore: ", err)
+		return ""
+	}
+	return ""
+}
+
+func (c *Client) ReleaseFairSemaphore(semname, identifier string) bool {
+	pipeline := c.Conn.TxPipeline()
+	pipeline.ZRem(semname, identifier)
+	pipeline.ZRem(semname+":owner", identifier)
+	if _, err := pipeline.Exec(); err != nil {
+		log.Println("pipeline err in ReleaseFairSemaphore: ", err)
+		return false
+	}
+	return true
+}
+
+func (c *Client) RefreshFairSemaphore(semname, identifier string) bool {
+	if c.Conn.ZAdd(semname, &redis.Z{Member: identifier, Score: float64(time.Now().Unix())}).Val() != 0 {
+		c.ReleaseFairSemaphore(semname, identifier)
+		return false
+	}
+	return true
+}
+
+func (c *Client) AcquireSemaphoreWithLock(semname string, limit int64, timeout int64) string {
+	identifier := c.AcquireLock(semname, 0.01)
+	if identifier != "" {
+		return c.AcquireFairSemaphore(semname, limit, timeout)
+	}
+	defer c.ReleaseLock(semname, identifier)
+	return ""
+}
+
+type soldData struct {
+	SellerId string
+	ItemId   string
+	Price    string
+	BuyerId  string
+	Time     int64
+}
+
+func (c *Client) SendSoldEmailViaQueue(seller, item, price, buyer string) {
+	data := soldData{
+		SellerId: seller,
+		ItemId:   item,
+		Price:    price,
+		BuyerId:  buyer,
+		Time:     time.Now().Unix(),
+	}
+	jsonValue, err := json.Marshal(data)
+	if err != nil {
+		log.Println("marshal err in SendSoldEmailViaQueue: ", err)
+		return
+	}
+	c.Conn.RPush("queue:email", jsonValue)
+}
+
+func (c *Client) ProcessSoldEmailQueue() {
+	for !common.QUIT {
+		packed := c.Conn.BLPop(30*time.Second, "queue:email").Val()
+		if len(packed) == 0 {
+			continue
+		}
+
+		toSend := soldData{}
+		if err := json.Unmarshal([]byte(packed[0]), &toSend); err != nil {
+			log.Println("unmarshal err in ProcessSoldEmailQueue: ", err)
+			return
+		}
+
+		SendEmail()
+	}
+}
+
+func SendEmail() {}
+
+type info struct {
+	Identifier string
+	Queue      string
+	Name       string
+	Args       []string
+}
+
+func (c *Client) ExecuteLater(queue, name string, args []string, delay float64) string {
+	identifier := uuid.NewV4().String()
+	data := info{
+		Identifier: identifier,
+		Queue:      queue,
+		Name:       name,
+		Args:       args,
+	}
+
+	item, err := json.Marshal(data)
+	if err != nil {
+		log.Println("marshal err in ExecuteLater: ", err)
+		return ""
+	}
+
+	if delay > 0 {
+		c.Conn.ZAdd("delayed:", &redis.Z{Member: item, Score: float64(time.Now().UnixNano() + int64(delay*1e9))})
+	} else {
+		c.Conn.RPush("queue:"+queue, item)
+	}
+	return identifier
+}
+
+func (c *Client) PollQueue(channel chan struct{}) {
+	for !common.QUIT {
+		item := c.Conn.ZRangeWithScores("delayed:", 0, 0).Val()
+		if len(item) == 0 || int64(item[0].Score) > time.Now().UnixNano() {
+			time.Sleep(100 * time.Millisecond)
```

Comment:
10ms in the book, mostly dependent on how much latency folks are willing to put up with. Folks will adjust as necessary if they use this. No worries :)

Reply:
Emm... Consistent with book content may reduce readers' confusion when they are reading. Fix it to 10ms, all test ok~